Source/JavaScriptCore/builtins/PromiseConstructor.js

```diff
@@ -78,6 +79,106 @@ function all(iterable)
     return promiseCapability.@promise;
 }
 
+@linkTimeConstant
+function promiseOnRejectedWithContext(argument, context)
+{
+    "use strict";
+
+    @rejectPromiseWithFirstResolvingFunctionCallCheck(context.globalContext.promise, argument);
+}
+
+@linkTimeConstant
+function promiseAllOnFulfilled(argument, context)
+{
+    "use strict";
+
+    var globalContext = context.globalContext;
+    var values = globalContext.values;
+
+    @putByValDirect(values, context.index, argument);
+
+    if (!--globalContext.remainingElementsCount)
+        @resolvePromiseWithFirstResolvingFunctionCallCheck(globalContext.promise, values);
+}
+
+@linkTimeConstant
+function promiseNewOnRejected(promise)
+{
+    "use strict";
+
+    return function @reject(reason) {
+        return @rejectPromiseWithFirstResolvingFunctionCallCheck(promise, reason);
+    };
+}
+
+@linkTimeConstant
+function promiseAllNewResolveElement(globalContext, index)
+{
+    "use strict";
+
+    var alreadyCalled = false;
+    return (argument) => {
+        if (alreadyCalled)
+            return @undefined;
+        alreadyCalled = true;
+
+        var values = globalContext.values;
+        @putByValDirect(values, index, argument);
+
+        if (!--globalContext.remainingElementsCount)
+            return @resolvePromiseWithFirstResolvingFunctionCallCheck(globalContext.promise, values);
+    };
+}
+
+function all(iterable)
+{
+    "use strict";
+
+    if (this !== @Promise)
+        return @promiseAllSlow.@call(this, iterable);
+
+    var promise = @newPromise();
+    var values = [];
+    var globalContext = {
+        promise,
+        values,
+        remainingElementsCount: 1,
+    };
+    var index = 0;
+    var onRejected;
+
+    try {
+        var promiseResolve = this.resolve;
+        if (!@isCallable(promiseResolve))
+            @throwTypeError("Promise resolve is not a function");
+
+        for (var value of iterable) {
+            @putByValDirect(values, index, @undefined);
+            var nextPromise = promiseResolve.@call(this, value);
+            ++globalContext.remainingElementsCount;
+            var then = nextPromise.then;
+            if (@isPromise(nextPromise) && then === @defaultPromiseThen) {
+                var constructor = @speciesConstructor(nextPromise, @Promise);
+                if (constructor !== @Promise)
+                    var promiseOrCapability = @newPromiseCapabilitySlow(constructor);
```

Comment:
I know we have to use `var`s in builtins JS, but I definitely don't think we should hoist them. 😅